src/optimize/base_optimizer.js

```diff
@@ -102,6 +107,56 @@ export default class BaseOptimizer {
     });
   }
 
+  warmupThreadLoaderPool() {
+    const baseModules = [
+      'babel-loader',
+      BABEL_PRESET_PATH
+    ];
+
+    const devOnlyModules = !IS_KIBANA_DISTRIBUTABLE
+      ? ['ts-loader']
+      : [];
+
+    threadLoader.warmup(
+      // pool options, like passed to loader options
+      // must match loader options to boot the correct pool
+      this.getThreadLoaderPoolConfig(),
+      [
+        // modules to load on the pool
+        ...baseModules,
+        ...devOnlyModules
+      ]
+    );
+  }
+
+  getThreadLoaderPoolConfig() {
+    // In some cases cpus() returns undefined
+    // https://github.com/nodejs/node/issues/19022
+    //
+    // The logic to calculate the cpus is the following:
+    // - In case we can't calculate the number of cpus we will define it
+    // to be 4
+    // - Otherwise we will use the available number - 1 except for the cases where
+    // we only have 1 cpu. For those cases the final calculated cpus will also be 1
+    const cpus = os.cpus() || { length: 0 };
+    const currentAvailableOsCpus =  Math.max(0, cpus.length - 1) || 1;
+    const calculatedCpus = currentAvailableOsCpus > 0 ? currentAvailableOsCpus : 4;
+
+    return {
+      name: 'optimizer-thread-loader-main-pool',
+      workers: calculatedCpus,
+      workerParallelJobs: 20,
+      // This is a safe check in order to set
+      // max_old_space_size in the workers
+      // accordingly to memory allowed in the main process.
+      // Otherwise, if the user sets max_old_space_size into
+      // NODE_OPTIONS, it won't affect the workers.
+      workerNodeArgs: [`--max_old_space_size=${Math.trunc(v8.getHeapStatistics().total_heap_size / 1000000)}`],
```

Comment:
thoughts on putting these (thread, heap) in kibana.yml? I think we need some care here not to take over an entire machine, so I'd push for cautious defaults and optional overrides.

Reply:
By default it will be applied the node defaults. In case a user defines the `--max_old_space_size` like they are doing right now in some cases, all the spawned workers will have a max limit equal to the one set by the user with `--max_old_space_size`. However in development this will always return the `Node.js` defaults as if we applied for example `node --max_old_space_size=4096 scripts/kibana --dev` this flag won't be propagated to the optimizer forked processes from the main kibana dev server process. Any thoughts on that @spalger @jbudz  ? Do you think this is a thing we should fix when spawn the kibana dev server?